toolchain/sem_ir/inst_namer.h

```diff
@@ -152,6 +152,58 @@ class InstNamer {
     Namespace labels;
   };
 
+  // Helper class for naming a single instruction.
+  class NamingContext {
+   public:
+    // Enqueues the provided instructions for naming.
+    using QueueBlockInstsFn = llvm::function_ref<
+        auto(ScopeId scope_id, llvm::ArrayRef<InstId> inst_ids)->void>;
+
+    explicit NamingContext(InstNamer* inst_namer,
+                           QueueBlockInstsFn queue_block_insts,
+                           InstNamer::ScopeId scope_id, InstId inst_id);
+
+    // Names the single instruction. Use bound names where available. Otherwise,
+    // assign a backup name.
+    //
+    // Insts with a type_id are required to add names; other insts may
+    // optionally set a name. All insts may enqueue other insts to be named.
+    auto NameInst() -> void;
+
+   private:
+    // Adds the instruction's name.
+    auto AddInstName(std::string name) -> void;
+
+    // Adds the instruction's name by `NameId`.
+    auto AddInstNameId(NameId name_id, llvm::StringRef suffix = "") -> void {
+      AddInstName(
+          (sem_ir().names().GetIRBaseName(name_id).str() + suffix).str());
+    }
+
+    // Names an `IntType` or `FloatType`.
+    auto AddIntOrFloatTypeName(char type_literal_prefix, InstId bit_width_id,
+                               llvm::StringRef suffix = "") -> void;
+
+    // Names an `ImplWitnessTable` instruction.
+    auto AddWitnessTableName(InstId witness_table_inst_id, std::string name)
+        -> void;
+
+    // Enqueues all instructions in a block, by ID.
+    auto QueueBlockId(ScopeId scope_id, InstBlockId block_id) -> void {
+      if (block_id.has_value()) {
+        queue_block_insts_(scope_id, sem_ir().inst_blocks().Get(block_id));
+      }
+    }
+
+    auto sem_ir() -> const SemIR::File& { return *inst_namer_->sem_ir_; }
+
+    InstNamer* inst_namer_;
+    QueueBlockInstsFn queue_block_insts_;
```

Comment:
The callback here seems avoidable, and I think it might be nice to do so. I wonder if we could instead pass a reference to the `insts` vector and make the `queue_block_insts` lambda a real function that takes a reference to the vector? Or something?

Reply:
Done: https://github.com/carbon-language/carbon-lang/pull/5503/commits/9546e7fcb523d67afee7722d696d6f74f1e1ed15